Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -1,443 +1,381 @@
-/*
-Original Author: Michael Ash on 11/9/08
-Copyright (c) 2008 Rogue Amoeba Software LLC
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Foundation/NSOperationQueue.h>
+
+#import <Foundation/NSBlockOperation.h>
+#import <Foundation/NSCondition.h>
 
-Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
-documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
-rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
-persons to whom the Software is furnished to do so, subject to the following conditions:
+#import "NSOperationInternal.h"
+#import "Starboard.h"
 
-The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
-Software.
+#import <atomic>
+#import <mutex>
+#import <queue>
+
+static const wchar_t* g_logTag = L"NSOperationQueue";
+
+// Private helper that converts NSQualityOfService -> QOS_CLASS
+static inline long _QOSClassForNSQualityOfService(NSQualityOfService quality) {
+    switch (quality) {
+        case NSQualityOfServiceUserInteractive:
+            return QOS_CLASS_USER_INTERACTIVE;
+        case NSQualityOfServiceUserInitiated:
+            return QOS_CLASS_USER_INITIATED;
+        case NSQualityOfServiceUtility:
+            return QOS_CLASS_UTILITY;
+        case NSQualityOfServiceBackground:
+            // currently generates a warning:
+            // 'QOS_CLASS_BACKGROUND' is deprecated: QOS_CLASS_BACKGROUND is the same as DISPATCH_QUEUE_PRIORITY_LOW on WinObjC
+            // should go away once libdispatch is updated
+            return QOS_CLASS_BACKGROUND;
```

Comment:
I think I'd prefer to disable the warning in this instance, since it's one that'll go away once we Do The Right Thing and update libdispatch.